### DISCUSSION TITLE
Start and stop menu?

### DISCUSSION THREAD

**Discussion by DisasterAreaDesigns** [davetcc/tcMenu#414] [Q&A] [answered]
Hi there!  I was wondering whether it was possible to start / stop the menu in an application?  Let's say that I want to have a device that turns on an LED when I press a button, but when I hold the button I want to start up the menu.  Then the menu would run until I select an "exit" item on the top level menu.  Can I just call taskManager.runLoop() when I want to run the menu?

Something like this:

```
byte runMyMenu = 0;

void loop {
if(myButtonHeld == 1) {
runMyMenu = 1;
}
while (runMyMenu) {
taskManager.runLoop();
}
}
```

When I select the "exit" item in the menu, I would then set runMyMenu = 0 again and the non-menu portion of the code would run.  Possible?  I would like to run a fairly timing-sensitive operation in the non-menu part of the code as well as showing something other than the menu on the display.  The menu would just be for configuration / setup of the device.

**Comment by davetcc** [comment]
It is possible to take over the display for your own purposes and give it back some point later: https://www.thecoderscorner.com/products/arduino-libraries/tc-menu/renderer-take-over-display/

**Reply by DisasterAreaDesigns** [reply]
Right, I get that.  But isn't that more for painting something to the display that's outside the menu?  In your example you're waiting for the callback and then drawing an icon.  I'm really talking about stopping the menu until I invoke it, or at least completely hiding it.  I don't know what kind of performance impact running the menu will have since there really isn't a "background" on a SAMD21.

**Reply by davetcc** [reply]
The overhead of the menu running is fairly low, depending on what plugins you've used the only things that should be running are a task to service the input, and one for display, I can't think of much else. When you take over the display the task for drawing is in your hands, so you can determine what it's going to do (maybe nothing).

In terms of completely disabling task manager, you can do that too, you can even selectively disable tasks, but that's a bit outside the scope of the normal use cases, although it should be possible. If you have any kind of remote capability built in, then it will have its own tasks, and you cannot stop task manager without stopping the remotes from working.

**Reply by DisasterAreaDesigns** [reply]
That sounds reasonable, thanks.  If the menu / taskmanager is going to be watching inputs for menu interaction, is it possible to hijack those same inputs for my non-menu tasks?  Mainly I'd want to know when a button was tapped / pressed or held.

Unrelated: is it possible to specify members of an array for variables used in the menu?  For example, if I wanted to have int myVar[4]; and then in the menu have the user able to change or view myVar[1] in one item and myVar[3] in another, for example?  I tried using myVar[0] as Variable Name in the menu designer and it throws a "some fields are preventing save" error.  I'd hate to have to read out of several dozen or hundred values to stuff into an array when it would be a lot easier to pass those variables into the array to start with.

**Reply by davetcc** [reply]
Please can you post new a discussion for each question in future, I didn't even notice the 2nd question.

Short answer, no you can't directly create the items in an array, they are in a tree, and that would be difficult to create code for that. However, they are already in a tree / linked list, so maybe you don't need to do that.

You could start at the submenu root item, and then call `sub.getChild()` to get the first, and the rest follow on from there using `item->getNext()` until it returns null. There are even inbuilt iterators for this in the `MenuIterator.h` header. There's a lot of documentation around how the menu items are stored. Probably best to read through the docs around this: https://www.thecoderscorner.com/products/arduino-libraries/tc-menu/menu-item-types/ but there's a lot of other docs around this too.

**Reply by davetcc** [reply]
also https://www.thecoderscorner.com/products/arduino-libraries/tc-menu/menumanager-and-iteration/#iterating-over-menu-items

**Reply by davetcc** [reply, marked as answer]
So along with the above answer, for anyone wanting more manual control, TcMenu now supports fluent API - https://www.thecoderscorner.com/products/arduino-libraries/tc-menu/menu-item-types/fluent-menu-builder-intro/, where you can take far more manual control of the menu items.
